### DISCUSSION TITLE
[GSoC] 9. Add ensembling methods for tiling to Anomalib

### DISCUSSION THREAD

**Discussion by samet-akcay** [openvinotoolkit/openvino#16267] [Google Summer of Code]
# Short description
Detecting small defects in high-resolution images can be a challenging task for anomaly detection models. Anomalib currently has a tiling mechanism in place to improve the detection capabilities for such datasets, which involves dividing the input images into a grid of tiles which are then processed separately by the model. A limitation of the tiling mechanism is that a single model is trained on all tile locations combined, leaving the approach ineffective for locally-aware models that require a fixed position and orientation of the objects in the images. For such models, an ensemble approach would be required. The idea is to train separate models for each of the tile locations and combine the predictions of the models in the post-processing stage. This project involves adding such an ensemble approach to tiling to the Anomalib library.

# Expected outcomes
Data-to-prediction pipeline incorporating ensemble-based tiling

# Skills required/preferred
ML basics, Python

# Mentors
Dick Ameln, Samet Akcay

# Size of project
175 hours

# Difficulty
Medium

**Comment by blaz-r** [comment]
Hello, 

I'm Blaž Rolih, a first-year master's student at the Faculty of Computer and Information science, University of Ljubljana in Slovenia. I'm interested in artificial intelligence, especially machine learning and deep learning in computer vision, but also enjoy working with embedded systems and writing software in general. 

I have contributed to some open-source projects before, and have some of my own. Since I like the idea of open source, I decided that I will contribute to a larger project, and GSoC seems like a really good start for this. 

I worked a bit with OpenVINO in my bachelor's thesis, where I implemented and evaluated gesture recognition from video, using deep learning on an embedded device. I haven't worked with Anomalib before, but it looks very useful and nice library, so I'd like to contribute as best as I can.

That's why I'm inquiring about this specific project of adding ensembling methods for tiling in Anomalib. I did some research at faculty about ensemble methods and how they can improve performance and stability of prediction, so it would be nice to implement and evaluate this in Anomalib as well. 

I have already completed the prerequisite task, where I converted Segmenter, the model for semantic segmentation, to OpenVINO.
Are there any recommended steps, besides getting familiar with Anomalib, that I should do to tackle this problem? It would also be nice if you could please help me construct the plan for implementation and proposal, so it covers all the parts and details that you had in mind.

Thank you for your time. 

Best regards,
Blaž

**Reply by adrianboguszewski** [reply]
@samet-akcay, could you advise on this?

**Comment by samet-akcay** [comment]
Hi Blaž, thanks for your interest! 

@djdameln, can you provide a more detailed description here?

**Reply by blaz-r** [reply]
Before the GSoC officially starts, I will go through some issues, that are open in the Anomalib repository. This way I can familiarize myself with the library and fix some bugs / solve some problems in the process.
If you think anything specific requires my attention, I'd be glad to check it out. I'll try to contribute as much as my time besides the studies currently allows.

**Comment by blaz-r** [comment]
Hello @samet-akcay @djdameln
Applications for GSoC open tomorrow and I'll apply for this project, should I send you a draft of the proposal to email so we can polish it, or do you prefer anything else?
I did some testing with anomalib and even opened some new issues and potentially fixed some. I find it well designed and I think it has great potential for additions like this ensemble method. I'm still not entirely familiar with the entire structure, and the tiling mechanism, so it'd be great to hear your input on this.

**Comment by samet-akcay** [comment]
Hi @blaz-r, thank you for being active on anomalib!
If you want to get some feedback on your proposal, you could email to either [me](samet.akcay@intel.com) or [@djdameln](dick.ameln@intel.com)

**Reply by blaz-r** [reply]
I sent proposal draft via email yesterday for a review. Thank you very much for your time and help in advance.

**Reply by samet-akcay** [reply]
Thanks @blaz-r, we are currently reviewing your proposal

**Comment by AmishKakka** [comment]
@samet-akcay 
I sent my proposal yesterday via mail. Please review it and advise some feedback too.

**Reply by samet-akcay** [reply]
Thanks @AmishKakka! Please allow us some time since we have recently received a number of proposals.

**Reply by samet-akcay** [reply]
@AmishKakka, can you please send your proposal again. I cannot find yours. Thanks!

**Reply by AmishKakka** [reply]
I have sent it again. Please check it. Thanks!

**Comment by AmishKakka** [comment]
Can you review my proposal again, I have updated it

**Reply by samet-akcay** [reply]
@AmishKakka, can you email it to me again please?
